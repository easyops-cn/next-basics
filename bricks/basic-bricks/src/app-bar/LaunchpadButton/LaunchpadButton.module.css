.launchpadLink {
  width: 24px;
  height: 24px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.launchpadIcon {
  vertical-align: top;
  filter: grayscale(100%);
  opacity: 0.8;
}

.launchpadLink:hover .launchpadIcon {
  color: var(--color-brand-hover) !important;
  filter: none;
  opacity: 1;
}
