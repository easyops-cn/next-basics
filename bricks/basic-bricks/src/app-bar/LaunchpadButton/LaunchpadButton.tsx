import React, { useCallback, useEffect, useState } from "react";
import { getRuntime, getHistory } from "@next-core/brick-kit";
import { ReactComponent as LaunchpadSvg } from "../../images/launchpad.svg";
import styles from "./LaunchpadButton.module.css";
import { LaunchpadPortal } from "../LaunchpadPortal/LaunchpadPortal";
import hotkeys from "hotkeys-js";
import { launchpadService } from "../LaunchpadService";

export function LaunchpadButton(): React.ReactElement {
  const [visible, setVisible] = useState(false);
  const openLaunchpad = (): void => {
    setVisible(true);
  };

  const handleLaunchpadClose = (): void => {
    setVisible(false);
  };
  const handleLaunchpadWillClose = useCallback((): void => {
    getRuntime().toggleLaunchpadEffect(false);
  }, []);

  useEffect(() => {
    launchpadService.init();
  }, []);

  useEffect(() => {
    const unlisten = getHistory().listen(() => {
      // 当切换页面时，关闭 Launchpad。
      handleLaunchpadWillClose();
      setVisible(false);
    });
    return unlisten;
  }, [handleLaunchpadWillClose]);

  useEffect(() => {
    hotkeys.filter = function () {
      return true;
    };

    hotkeys("alt+l", (event) => {
      event.preventDefault();
      setVisible((preVisble) => {
        const curVisble = !preVisble;
        if (!curVisble) handleLaunchpadWillClose();
        return curVisble;
      });
    });

    return () => {
      hotkeys.unbind("alt+l");
    };
  }, [handleLaunchpadWillClose]);

  useEffect(() => {
    // 当打开/关闭 Launchpad 时，切换背景模糊，切换悬浮展开模式；
    getRuntime().toggleLaunchpadEffect(visible);
  }, [visible]);

  return (
    <>
      <a role="button" className={styles.launchpadLink} onClick={openLaunchpad}>
        <LaunchpadSvg className={styles.launchpadIcon} width={16} height={16} />
      </a>
      <LaunchpadPortal
        visible={visible}
        onWillClose={handleLaunchpadWillClose}
        onClose={handleLaunchpadClose}
      />
    </>
  );
}
