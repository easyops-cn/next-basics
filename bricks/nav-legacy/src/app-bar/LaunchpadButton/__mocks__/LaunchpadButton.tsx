import React from "react";

export function LaunchpadButton(): React.ReactElement {
  return <div>LaunchpadButton</div>;
}
