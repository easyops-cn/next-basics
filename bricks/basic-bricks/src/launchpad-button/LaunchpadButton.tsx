import React from "react";
import * as originLanuchpad from "../app-bar/LaunchpadButton/LaunchpadButton";

export function LaunchpadButton(): React.ReactElement {
  return <originLanuchpad.LaunchpadButton />;
}
