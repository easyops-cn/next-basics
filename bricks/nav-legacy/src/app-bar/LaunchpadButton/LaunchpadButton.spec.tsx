import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import { getRuntime, getHistory } from "@next-core/brick-kit";
import { LaunchpadButton } from "./LaunchpadButton";
import {
  LaunchpadPortal,
  LaunchpadPortalProps,
} from "../LaunchpadPortal/LaunchpadPortal";

jest.mock("@next-core/brick-kit");

type ListenerFn = () => void;

const listeners = new Set<ListenerFn>();
const spyOnHistoryUnListen = jest.fn();
const spyOnHistoryListen = jest.fn((listener: ListenerFn): (() => void) => {
  listeners.add(listener);

  return () => {
    listeners.delete(listener);
    spyOnHistoryUnListen();
  };
});
(getHistory as jest.Mock).mockReturnValue({
  listen: spyOnHistoryListen,
});

const spyOnToggleLaunchpadEffect = jest.fn();
const spyOnIsMenuBarCollapsed = jest.fn();
const spyOnIsMenuBarSoftExpanded = jest.fn();
(getRuntime as jest.Mock).mockReturnValue({
  menuBar: {
    isCollapsed: spyOnIsMenuBarCollapsed,
    isSoftExpanded: spyOnIsMenuBarSoftExpanded,
  },
  toggleLaunchpadEffect: spyOnToggleLaunchpadEffect,
});
jest.mock("../LaunchpadService", () => {
  return {
    launchpadService: {
      fetchFavoriteList: () => [],
    },
  };
});

describe("LaunchpadButton", () => {
  let wrapper: ShallowWrapper;
  const getPortal = (): ShallowWrapper<LaunchpadPortalProps> =>
    wrapper.find(LaunchpadPortal);

  beforeEach(() => {
    wrapper = shallow(<LaunchpadButton />);
  });

  afterEach(() => {
    spyOnToggleLaunchpadEffect.mockClear();
  });

  it("should work", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should open launchpad", async () => {
    //await (global as any).flushPromises();
    expect(getPortal().prop("visible")).toBe(false);
    wrapper.find("a").simulate("click");
    await (global as any).flushPromises();
    expect(getPortal().prop("visible")).toBe(true);
  });

  it("should toggle filter of blur to false", () => {
    getPortal().invoke("onWillClose")();
    expect(spyOnToggleLaunchpadEffect).toBeCalledWith(false);
  });

  it("should set visible to false", () => {
    getPortal().invoke("onClose")();
    expect(getPortal().prop("visible")).toBe(false);
  });
});
