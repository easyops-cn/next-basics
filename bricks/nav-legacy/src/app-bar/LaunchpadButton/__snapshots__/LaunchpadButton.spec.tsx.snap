// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchpadButton should work 1`] = `
<Fragment>
  <a
    className="launchpadLink"
    onClick={[Function]}
    role="button"
  >
    <svgr-mock
      className="launchpadIcon"
      height={16}
      width={16}
    />
  </a>
  <LaunchpadPortal
    onClose={[Function]}
    onWillClose={[Function]}
    visible={false}
  />
</Fragment>
`;
